### ISSUE TITLE
function call that returns `Never` incorrectly marked as unreachable when function that's defined later with multiple argument decorator is called

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#13043]
sorry for the wacky title
```py
from typing import TypeVar, Literal
from typing_extensions import assert_never

Fn = TypeVar("Fn")


def foo(fn: Fn, value: bool = False) -> Fn:
    ...


def bar(value: Literal[True]) -> None:
    if value:
        baz()
    else:
        assert_never(value) # error: Statement is unreachable  [unreachable]


@foo
def baz() -> None:
    ...
```
[playground](https://mypy-play.net/?mypy=latest&python=3.10&flags=show-error-codes%2Cwarn-unreachable&gist=5d9c29537790b7eeeb72c88547869f0a)

functions that return `Never` are not supposed to be marked as unreachable - see https://github.com/python/mypy/issues/10916#issuecomment-1073526572

**Comment by KotlinIsland** [comment]
Looks similar to #8900

minified:
```py
from typing import _T as T

def bar() -> None:
    return baz()
    exit()

def foo(fn: T, a: bool = False) -> T: ...

@foo
def baz() -> None: ...
```

**Comment by DetachHead** [comment]
also causes an invalid 'Expression has type "Any"' error:
```py
from typing import TypeVar, Literal
from typing_extensions import assert_never

Fn = TypeVar("Fn")

def foo(fn: Fn, value: bool = False) -> Fn:
    ...


def f() -> None:
    baz()
    try:
        ...
    except Exception as ex:  # error: Expression has type "Any"  [misc]
        pass

@foo
def baz() -> None:
    ...
```
[playground](https://mypy-play.net/?mypy=latest&python=3.10&gist=013b49fd8859a4818ae2969dc44420b4&flags=strict%2Cdisallow-any-expr)

very related (probably the same issue?): #13167
